### ISSUE TITLE
jnp.linalg.inv is surprisingly slow

### ISSUE THREAD

**Issue by tbenthompson** [google/jax#11321]
I would expect `jnp.linalg.inv` to be fast. Instead, I'm able to beat it by a factor of ~30x by writing my own JAX explicit 3x3 inverse. The `jnp.linalg.inv` call below takes 1.47ms to do 10,000 3x3 inverses, whereas the hand-written version takes 0.041ms to do the same. 

There are ~50 flops required to compute a 3x3 matrix inverse. And the calculation uses 36 (9 floats * 4 bytes/float) bytes of memory. As a result, I'd expect the calculation to be slightly bandwidth-bound, but still running around 50ns per inverse on a single core, or 5ns per inverse with perfect scaling on 10 cores. This matches very closely with the 41us for 10,000 hand-written inversions on my laptop. 

The performance of `jnp.linalg.inv` on large matrices seems to be fine, so the problem here seems isolated to batch inversion of many small matrices. 

I've isolated this performance issue from a broader project where the inversion is currently taking about 90% of my runtime. With the handwritten version, I get that down to 5% or so. But, I don't want to have to roll my own matrix inversion routines for n = 1,2,3,4,5,6... It would be really nice to have high-performance batch inversion of small matrices!

The benchmark I've shown below is on my local laptop, a Mac M1. I've replicated the results on x64 CPU and Nvidia GPU on Colab. For the GPU benchmark in the screenshot, I increased the number of matrices to 100k instead of 10k since there seems to be some fixed overhead.  

```python
import numpy as np
import jax
import jax.numpy as jnp

def inv33(mat):
    m1,m2,m3 = mat[0]
    m4,m5,m6 = mat[1]
    m7,m8,m9 = mat[2]
    det = m1*m5*m9 + m4*m8*m3 + m7*m2*m6 - m1*m6*m8 - m3*m5*m7 - m2*m4*m9  
    inv_det = 1.0 / det
    return jnp.array([[m5*m9-m6*m8, m3*m8-m2*m9, m2*m6-m3*m5],
                     [m6*m7-m4*m9, m1*m9-m3*m7, m3*m4-m1*m6],
                     [m4*m8-m5*m7, m2*m7-m1*m8, m1*m5-m2*m4]]) * inv_det
mats = np.random.rand(10000, 3, 3)
jax_mats = jnp.array(mats)
vmap_inv33 = jax.jit(jax.vmap(inv33))
```

```
%%timeit -n 1000
jnp.linalg.inv(jax_mats).block_until_ready()
```
```
1.47 ms ± 13.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

```
%%timeit -n 1000
vmap_inv33(jax_mats).block_until_ready()
```
```
41 µs ± 899 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

<img width="749" alt="image" src="https://user-images.githubusercontent.com/4241811/176683207-f2f620f5-f4b0-4c51-b688-f06b06c132bf.png">

**Comment by tbenthompson** [comment]
I'm seeing the same large performance differential on a Colab TPU. Though note that I changed the handwritten algorithm slightly to pack the matrix into a 9 element vector. This seems to be much faster on the TPU. My guess is that it's because of the 128x8 packing, but I'm no TPU expert.

<img width="708" alt="image" src="https://user-images.githubusercontent.com/4241811/176685107-39303378-330f-453f-bb82-ebd0fb5e338a.png">

**Comment by YouJiacheng** [comment]
`inv` is backed by batched LU decomposition and batched triangular solve.
```python
from contextlib import contextmanager
@contextmanager
def timer():
    from time import perf_counter
    t = perf_counter()
    yield
    print(perf_counter() - t)

import jax
import jax.numpy as jnp
import numpy as np

x = jnp.array(np.random.rand(100000, 3, 3))
e = jnp.array(np.zeros((100000, 3, 3)) + np.eye(3))

jax.numpy.linalg.inv(x).block_until_ready()
with timer(): # 1.45s
    for _ in range(1000):
        jax.numpy.linalg.inv(x).block_until_ready()

jax.numpy.linalg.solve(x, e).block_until_ready()
with timer(): # 1.44s
    for _ in range(1000):
        jax.numpy.linalg.solve(x, e).block_until_ready()

@jax.jit
def mock_solve(a, b): # mock lu solve without lu factorization
    x = jax.lax.linalg.triangular_solve(a, b, left_side=True, lower=True, unit_diagonal=True)
    return jax.lax.linalg.triangular_solve(a, x, left_side=True, lower=False)

mock_solve(x, e).block_until_ready()
with timer(): # 1.21s
    for _ in range(1000):
        mock_solve(x, e).block_until_ready()

print(mock_solve.lower(x, e).compile().compiler_ir()[0].to_string()) # just call cublas_trsm_batched
```
It seems the main problem is `cublas_trsm_batched` is slow(~1200us per loop).
(`cublas_getrf_batched` is also slow(~200us per loop))

**Comment by froystig** [comment]
Thanks for filing!

Relatedly, we've proposed unrolling various linalg operations for small-size inputs (#4258). That would likely include writing an explicit handling of 3x3 matrix solves. Doing so would have masked this issue, for better and worse.

**Comment by tbenthompson** [comment]
In case it might be helpful, here's a file with the small JAX linear algebra operations that I've written. It's not the most beautiful thing (undocumented!) and there's some code generation happening but there's:
- determinant-based inverses for 2x2, 3x3, 4x4 matrices.
- recursion-based inverses for larger sizes with `inv_recurse`. These remain about 10x faster than `jnp.linalg.solve` all the way up to 20x20. The recursion uses block 2x2 inversion to reduce until there's a 4x4 or 3x3 or 2x2 matrix base case that uses the above determinant-based inversion.
- code generation to produce explicit, unrolled LU-based operations. These are really fast and a bit more numerically stable than the determinant inverses above. But there's still no pivoting so there's potential to get bad numerical behavior with ugly matrices. My use-case produces well-behaved matrices, but a general implementation would definitely need some upgrades.
  - LU decomposition
  - inverse from LU decomposition
  - linear solve from LU decomp

https://gist.github.com/tbenthompson/faae311ec4e465b0ff47b4aabe0d56b2

**Comment by EelcoHoogendoorn** [comment]
@tbenthompson thanks thats super useful.

I guess one might want to specialize the 66 case as a split in blocks of 3. 2 calls to the inv33 seems lower flop than a 4+2 call; plus I suppose the errors might be distributed more equally.

But then what about a 9x9? 3+(3+3) sure is cheaper than 4+4+1 as well? 

And 12x12?

Im guessing 12x12 is better done as (3+3)+(3+3) rather than (4+4)+4.

But I notice that with more recursion steps the numerical accuracy for A * inv(A) - I goes down a little.  So it isnt the case that using smaller blocks is always a pareto improvement. But in the intermediate scale where we shouldnt just revert to linalg.inv its an interesting question.

**Comment by EelcoHoogendoorn** [comment]
@tbenthompson 

Those unrolled code generated routines also look really cool; but im unable to get them to produce a working unit test.

The code looks like it is in a somewhat early state; are you aware of a newer version thats under test being somewhere out there?

**Comment by tbenthompson** [comment]
Unfortunately I've moved on and I'm not using `jax` in my day-to-day work anymore. Sorry to not be able to help more!

**Comment by EelcoHoogendoorn** [comment]
My condolences :). But its very helpful as-is anyway; should be part of jax or some other lib imo; perhaps ill get around to making it mergable somewhere. 

For general interest, on a recent jax version the solve is still about 10x faster than the builtin solve; and loses about 1 oom in precision, I suppose because of the lack of pivoting, but its a trade I will very happily take anyway. Happy to say that absolute performance is also very decent; comparing the number of flops to what a single core of mine should be able to do I get about 1 flop every 3 cycles; which really isnt shabby imo, for code like this that has very little opportunity for simd and is jumping around a lot. Ive seen a lot worse in the past for jax compiled code, but the compiler also has been improving I imagine.

**Comment by yhtang** [comment]
Just wanna add a datapoint that on B100, the manual inv is 20x faster when solving 1000000 3x3 systems.
```
>>> timeit.timeit('jnp.linalg.inv(jax_mats).block_until_ready()', number=1000, globals=globals())
1.4917597960156854
>>> timeit.timeit('vmap_inv33(jax_mats).block_until_ready()', number=1000, globals=globals())
0.0737372330040671
```
